### PR TITLE
Add automated image builds for VM, bare metal (amd64) and RPi (arm64-rpi)

### PR DESCRIPTION
**Pull request by openoms**
Images are so far in the successful builds at: https://github.com/openoms/raspiblitz/actions

Documentation: https://github.com/openoms/raspiblitz/blob/add-amd64-image-build/ci/README.md

Issue: #3053 and #2842

### CODE REVIEW

**Comment by openoms** [comment]
@rootzoll will change this to use the local `build_sdcard.sh` to test changes to it in the PRs.

To test the changes of the other scripts called will need to set the `build_sdcard.sh -b` option to use the PR branch - this is slightly more tricky to autodetect.

**Comment by openoms** [comment]
When this is merged will test all the PRs and commits to dev branch to make sure the build_sdcard.sh and the scripts called from it are functional.

**Comment by openoms** [comment]
arm64-rpi build and cleaner naming is in the works in https://github.com/openoms/raspiblitz/pull/74

**Comment by openoms** [comment]
This merge [d5569b8](https://github.com/rootzoll/raspiblitz/pull/3486/commits/d5569b8c507db1140813a96d3477217e514559dc) makes the github actions script assume that the repo owner is `rootzoll` so failed with the build.
A rebase could have worked instead, but no problem, merging in https://github.com/openoms/raspiblitz/pull/74 with a lot of improvements and the arm64-rpi build.

**Review by rootzoll — APPROVED** [review]

**Comment by openoms** [comment]
Rebased to the current `dev` branch.

**Comment by openoms** [comment]
now the rpi image uses the manually verified hash of the base image.

to sort out the local generation of the rpi image I am testing here to not trigger the ci continuously: https://github.com/openoms/raspiblitz/pull/75

**Comment by tlindi** [comment]
non-UEFI (aka Legacy BIOS) is quite legacy now days :-/

Do you have any plans to support UEFI images for amd64?

**Comment by openoms** [comment]
> non-UEFI (aka Legacy BIOS) is quite legacy now days :-/
> 
> Do you have any plans to support UEFI images for amd64?

most computers I checked have a non-UEFI boot option and I am not sure if you can install linux on the ones which don't have.
If can be done it could be a nice a feature, but outside of the scope of this PR.

**Comment by openoms** [comment]
Successfully generated an arm64-rpi-lean-image locally on Ubuntu Live (jammy). Extended the readme with the steps.
Ready to be merged.

**Comment by rootzoll** [comment]
@openoms `amd64-fatpack-image-build`failed ... is there still sometihng to so or how to retrigger this?

**Comment by openoms** [comment]
> @openoms `amd64-fatpack-image-build`failed ... is there still sometihng to so or how to retrigger this?

It can be rerun manually with the same settings for example if there was some temporary connection issue.
Any new commit (or PR once merged) will also trigger it. If the PR affects files which are ignored can also be triggered manually on the latest commit.

For now I merged the new commits from dev as a rebase would have been complicated.

**Comment by rootzoll** [comment]
Both fatpacks fail ... the amd64 with a ´segmentatin fault´during mempool install - is this because of the size of the sd card that now gets too small to have all those fatpack preinstalled apps?

```
    qemu:   }(global || this));
    qemu: generated-config.js file saved
==> qemu: Segmentation fault
    qemu: FAIL - npm run build did not run correctly, aborting (1)
Error: /14 14:42:25 packer-builder-qemu plugin: [ERROR] Remote command exited with '1': echo 'raspiblitz' | HOME_DIR='/home/pi' PACKER_BUILDER_TYPE='qemu' PACKER_BUILD_NAME='qemu' PACKER_HTTP_ADDR='10.0.2.2:8787' PACKER_HTTP_IP='10.0.2.2' PACKER_HTTP_PORT='8787' branch='add-amd64-image-build' github_user='openoms' http_proxy='' https_proxy='' no_proxy='' pack='fatpack'  sudo -S -E sh -eux '/tmp/script_3347.sh'
2022/12/14 14:42:25 packer-builder-qemu plugin: [INFO] RPC endpoint: Communicator ended with: 1
2022/12/14 14:42:25 [INFO] 1428018 bytes written for 'stdout'
2022/12/14 14:42:25 [INFO] 669524 bytes written for 'stderr'
2022/12/14 14:42:25 [INFO] RPC client: Communicator ended with: 1
2022/12/14 14:42:25 [INFO] RPC endpoint: Communicator ended with: 1
2022/12/14 14:42:25 packer-provisioner-shell plugin: [INFO] 1428018 bytes written for 'stdout'
2022/12/14 14:42:25 packer-provisioner-shell plugin: [INFO] 669524 bytes written for 'stderr'
2022/12/14 14:42:25 packer-provisioner-shell plugin: [INF
```

**Comment by openoms** [comment]
> Both fatpacks fail ... the amd64 with a ´segmentatin fault´during mempool install - is this because of the size of the sd card that now gets too small to have all those fatpack preinstalled apps?

Unlikely that it is a disk space limitation as the amd64 build is at 30GB and the arm64-rpi is set to 16GB. It is rather a mempool install issue, I think we should test it if it finished without mempool being built and/or run the build locally.

Is this closed by accident btw @rootzoll ?

**Comment by rootzoll** [comment]
I closed all "final testing" issues to get a better overview what is still to be worked on.
Just wanted todo it for issues - but also PRs got selected sorry - reopening.

**Comment by openoms** [comment]
I think the mempool install is jus ttoo much for the VM in github actions. It runs out of memory in https://github.com/openoms/raspiblitz/actions/runs/3705332669/jobs/6279104602 for amd64 after 5h27mins:
```
==> qemu: <--- Last few GCs --->
==> qemu:
==> qemu: [78497:0x5950760]   442125 ms: Mark-sweep 476.7 (499.8) -> 472.6 (500.0) MB, 2526.5 / 0.0 ms  (average mu = 0.277, current mu = 0.236) allocation failure; scavenge might not succeed
==> qemu: [78497:0x5950760]   446653 ms: Mark-sweep 476.7 (500.0) -> 472.5 (500.0) MB, 3768.9 / 0.0 ms  (average mu = 0.216, current mu = 0.168) allocation failure; scavenge might not succeed
==> qemu:
==> qemu:
==> qemu: <--- JS stacktrace --->
==> qemu:
==> qemu: FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
==> qemu:  1: 0xb6e500 node::Abort() [ng build --configuration production --localize]
==> qemu:  2: 0xa7e632  [ng build --configuration production --localize]
==> qemu:  3: 0xd47ea0 v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [ng build --configuration production --localize]
==> qemu:  4: 0xd48247 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [ng build --configuration production --localize]
==> qemu:  5: 0xf25605  [ng build --configuration production --localize]
==> qemu:  6: 0xf26508 v8::internal::Heap::RecomputeLimits(v8::internal::GarbageCollector) [ng build --configuration production --localize]
==> qemu:  7: 0xf36a13  [ng build --configuration production --localize]
==> qemu:  8: 0xf37888 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [ng build --configuration production --localize]
==> qemu:  9: 0xf121ee v8::internal::HeapAllocator::AllocateRawWithLightRetrySlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [ng build --configuration production --localize]
==> qemu: 10: 0xf135b7 v8::internal::HeapAllocator::AllocateRawWithRetryOrFailSlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [ng build --configuration production --localize]
==> qemu: 11: 0xef478a v8::internal::Factory::NewFillerObject(int, v8::internal::AllocationAlignment, v8::internal::AllocationType, v8::internal::AllocationOrigin) [ng build --configuration production --localize]
==> qemu: 12: 0x12b7d2f v8::internal::Runtime_AllocateInYoungGeneration(int, unsigned long*, v8::internal::Isolate*) [ng build --configuration production --localize]
==> qemu: 13: 0x16e9979  [ng build --configuration production --localize]
==> qemu: Aborted
    qemu: FAIL - npm run build did not run correctly, aborting (1)
```
For arm64-rpi it just ran out of the permitted 6h.

If the local run is ok there could be third option to run in github actions which which skips  the mempool and possible other install to keep under the 6h mark and other limitations.

**Comment by openoms** [comment]
I suggest we could merge this to use the Actions to test from the dev branch and from the PRs.
We could restrict it to pick up changes only in the build_sdcard.sh and the scripts being called from there (+ there is a manual trigger option) so it does not run on every commit when there is no change to what is running.
Rootzoll your merge will fail as it looks for the repo owner by the last user which committed to the branch.

File: ci/arm64-rpi/packer.build.arm64-rpi.sh [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  branch=$3
+else
+  branch=dev
+fi
+
+# Build the image in docker
+echo -e "\nBuild Packer image..."
+# from https://hub.docker.com/r/mkaczanowski/packer-builder-arm/tags
+docker run --rm --privileged -v /dev:/dev -v ${PWD}:/build \
+ mkaczanowski/packer-builder-arm:1.0.4@sha256:df09a8e249a292f10ca9b8cfd73420f5b987b6ac337d4ef28b6f4a8e61118822 \
```
Comment (dbast) [inline ci/arm64-rpi/packer.build.arm64-rpi.sh]: 1.0.5 has the qemu integration fixed https://github.com/mkaczanowski/packer-builder-arm/releases/tag/v1.0.5 (does not require the `setup-qemu-action` to run before and also works locally)
```suggestion
 mkaczanowski/packer-builder-arm:1.0.5@sha256:a6371a3230b94c308e9a8cd55da1ae0d23bd670b7e558b39683c24f489d03e53 \
```

Comment (openoms) [reply ci/arm64-rpi/packer.build.arm64-rpi.sh]: That is great, thank you for looking through this. Locally we actually prefer to not run in Docker so there is one less dependency, but noted that could run the image generation in Docker as well now. Could be a good option on a non-clean system - like an existing OS not started for specifically for compiling.

**Comment by openoms** [comment]
as the fatpack images are consistently failing I think we could disable the generation and keep only the lean images for now.

For example this error:
```
sudo: effective uid is not 0, is /usr/bin/sudo on a file system with the 'nosuid' option set or an NFS file system without root privileges?
```
is related to running in Docker and not straightforward to solve: https://github.com/docker/for-linux/issues/1100

**Comment by tlindi** [comment]
What is missing from lean vs fatpack - could there be readme instructions how to add those components after deploying Image?

**Comment by openoms** [comment]
> What is missing from lean vs fatpack - could there be readme instructions how to add those components after deploying Image?

Both have the same functionality except the WebUI and API is not active by default in the lean images. In fatpack the readily used binaries (like bitcoin core, LND) are downloaded and many things are precompiled like CLN, RTL, mempool, joinmarket etc, so there is less wait when activating them. 
Less of an issue on amd64 I'd say.

**Pull request merged**